File: vps_runner/presets.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Preset:
    name: str
    command: str
    timeout_sec: int = 180
    login_shell: bool = True


DEFAULT_APP_DIR = "/var/www/bitrix24-billing-app"
DEFAULT_BRANCH = "master"
DEFAULT_PM2 = "billing"


def billing_presets(
    app_dir: str = DEFAULT_APP_DIR,
    branch: str = DEFAULT_BRANCH,
    pm2_name: str = DEFAULT_PM2,
) -> list[Preset]:
    app = (app_dir or DEFAULT_APP_DIR).rstrip("/")
    branch = branch or DEFAULT_BRANCH
    pm2 = pm2_name or DEFAULT_PM2
    return [
        Preset(
            "Deploy billing",
            f"cd {app} && git pull origin {branch} && pm2 restart {pm2}",
            timeout_sec=300,
        ),
        Preset(
            "Git pull",
            f"cd {app} && git pull origin {branch}",
            timeout_sec=180,
        ),
        Preset("PM2 restart", f"pm2 restart {pm2}", timeout_sec=60),
        Preset("PM2 status", "pm2 status", timeout_sec=30),
        Preset(
            "PM2 logs",
            f"pm2 logs {pm2} --lines 120 --nostream",
            timeout_sec=30,
        ),
        Preset(
            "Git status",
            f"cd {app} && git status -sb && echo && git log -8 --oneline",
            timeout_sec=30,
        ),
    ]


def server_presets() -> list[Preset]:
    return [
        Preset(
            "Состояние сервера",
            "hostname; date; uptime; echo; df -hT; echo; free -h",
            timeout_sec=30,
        ),
        Preset(
            "Nginx reload",
            "nginx -t && (systemctl reload nginx || service nginx reload)",
            timeout_sec=30,
        ),
        Preset(
            "Nginx status",
            "systemctl status nginx --no-pager -l || service nginx status",
            timeout_sec=30,
        ),
    ]


def all_presets(
    app_dir: str = DEFAULT_APP_DIR,
    branch: str = DEFAULT_BRANCH,
    pm2_name: str = DEFAULT_PM2,
    include_server: bool = True,
) -> list[Preset]:
    items = billing_presets(app_dir, branch, pm2_name)
    if include_server:
        items = items + server_presets()
    return items

File: src/core/presets.cpp
#include "core/presets.hpp"

namespace fatty {

std::vector<Preset> deploy_presets(std::string app, std::string branch, std::string pm2) {
  if (app.empty()) app = kDefaultAppDir;
  while (!app.empty() && app.back() == '/') app.pop_back();
  if (branch.empty()) branch = kDefaultBranch;
  if (pm2.empty()) pm2 = kDefaultPm2;
  return {
      {"Deploy", "cd " + app + " && git pull origin " + branch + " && pm2 restart " + pm2, 300, true,
       "git pull и перезапуск pm2"},
      {"Git pull", "cd " + app + " && git pull origin " + branch, 180, true, "подтянуть ветку без перезапуска"},
      {"PM2 restart", "pm2 restart " + pm2, 60, true, "перезапустить процесс"},
      {"PM2 status", "pm2 status", 30, true, "список процессов"},
      {"PM2 logs", "pm2 logs " + pm2 + " --lines 120 --nostream", 30, true, "последние 120 строк, без follow"},
      {"Git status", "cd " + app + " && git status -sb && echo && git log -8 --oneline", 30, true,
       "ветка и последние коммиты"},
  };
}

std::vector<Preset> server_presets() {
  return {
      {"Состояние сервера", "hostname; date; uptime; echo; df -hT; echo; free -h", 30, true, "диск, память, uptime"},
      {"Nginx reload", "nginx -t && (systemctl reload nginx || service nginx reload)", 30, true,
       "проверка конфига и reload"},
      {"Nginx status", "systemctl status nginx --no-pager -l || service nginx status", 30, true, "статус сервиса"},
  };
}

std::vector<Preset> all_presets(const std::string& app_dir, const std::string& branch, const std::string& pm2_name,
                                bool include_server) {
  auto items = deploy_presets(app_dir, branch, pm2_name);
  if (include_server) {
    auto extra = server_presets();
    items.insert(items.end(), extra.begin(), extra.end());
  }
  return items;
}

}  // namespace fatty
